### PR TITLE
Completer

### PR DESCRIPTION
**Pull request by champs794**

### CODE REVIEW

File: practice/completer/site.js [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+   * They wanted it gone. It does not matter if it existed.
+   *
+   * @param {string} str
+   */
+  removeCompletion: function(str) {
+    var index = this.completions.indexOf(str);
+    if (index >= 0) {
+      this.completions.splice(index, 1);
+    }
+  },
```
Comment (stephenmuller) [inline practice/completer/site.js]: the lodash pull function will mutate the array without the need for an index

File: practice/completer/site.js [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+  /**
+   * Wrapper for document.getElementsByTagName()
+   *
+   * @param {string} tag_name
+   * @return NodeList
+   */
+  this.name = function(tagName) {
+    return document.getElementsByTagName(tagName);
+  };
+}
```
Comment (stephenmuller) [inline practice/completer/site.js]: I'm not exactly clear what the wrappers are for, do they just alias functions from other libraries?

Comment (champs794) [inline practice/completer/site.js]: Yeah, this was just a shortcut to some of its functionality. I'm very used to jQuery, but it hadn't been introduced.

File: practice/completer/site.js [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+    'nicollet',
+    'lyndale',
+    'portland',
+    'oakland'
+  ];
+
+  var rando = Math.random() * candidates.length;
+  $.id('item').value = candidates[Math.floor(rando)];
+}
+
```
Comment (stephenmuller) [inline practice/completer/site.js]: What does this do?

File: practice/completer/site.js [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+/**
+ * removes the selected text from the list
+ */
+function removeSelection() {
+  var text = document.getSelection();
+  $.id('item').value = text;
+  removeItem();
+}
```
Comment (stephenmuller) [inline practice/completer/site.js]: Is this specifically for removing text via the html 'website'?

File: practice/completer/site.js [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+   */
+  getFuzzyExpression: function(text) {
+    if (!this.expressions[text]) {
+      // kill nonword letters
+      var ret = text.replace(/\W/g,'');
+      // allow letters to squeeze in between
+      ret = ret.split('').join('.*');
+      var re = new RegExp(ret, 'i');
+      this.expressions[text] = re;
+    }
```
Comment (stephenmuller) [inline practice/completer/site.js]: re and ret made this tricky-ish to follow

File: practice/completer/site.js (hunk 1 above)
Comment (champs794) [inline practice/completer/site.js]: Nice.

File: practice/completer/site.js (hunk 3 above)
Comment (champs794) [inline practice/completer/site.js]: When you double click the label/box (I forget which), you populate the input with a random-ish value that isn't in the default list.

File: practice/completer/site.js (hunk 4 above)
Comment (champs794) [inline practice/completer/site.js]: Double clicking a word selects the entire word. The process from here:

1. Capture the event
1. Put the selected text in the input box
1. Trigger the remove button click event

…and then you can quickly put the word back into the list.

File: practice/completer/site.js (hunk 5 above)
Comment (champs794) [inline practice/completer/site.js]: I would not disagree.

File: practice/completer/site.js [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+    'witham',
+    'woodstock',
+    'woodward',
+    'yamhill',
+    'yeon',
+  ];
+
+  _.forEach(CITY_AVES, function(ave) {
+    genius.addCompletion(ave);
+  });
```
Comment (selassid) [inline practice/completer/site.js]: Useful trick: if your anonymous function is just calling another function, just pass that directly as the first function.

```js
_.forEach(CITY_AVES, genius.addCompletion);
```

File: practice/completer/site.js [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+   */
+  this.name = function(tagName) {
+    return document.getElementsByTagName(tagName);
+  };
+}
+
+/**
+ * Completer object methods
+ */
+var CompleterProto = {
```
Comment (selassid) [inline practice/completer/site.js]: Only the constructors should be in WordCase. All the other variables should be in camelCase.

File: practice/completer/site.js [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+   */
+  getFuzzyExpression: function(text) {
+    if (!this.expressions[text]) {
+      // kill nonword letters
+      var pattern = text.replace(/\W/g,'');
+      // allow letters to squeeze in between
+      pattern = pattern.split('').join('.*');
+      var re = new RegExp(pattern, 'i');
+      this.expressions[text] = re;
+    }
```
Comment (selassid) [inline practice/completer/site.js]: Cool idea and use of memoization.

**Pull request merged**